src/nvim/message.c

```diff
@@ -3064,14 +3077,24 @@ int msg_end(void)
 
 void msg_ext_ui_flush(void)
 {
-  if (!ui_has(kUIMessages)) {
+  if (!msg_enable_ext()) {
     return;
   }
 
   msg_ext_emit_chunk();
   if (msg_ext_chunks.size > 0) {
-    ui_call_msg_show(cstr_to_string(msg_ext_kind),
-                     msg_ext_chunks, msg_ext_overwrite);
+    if (ui_has(kUIMessages)) {
+      ui_call_msg_show(cstr_to_string(msg_ext_kind),
+                       msg_ext_chunks, msg_ext_overwrite);
+    } else {
+      msg_call_msgfunc("msg_show", &msg_ext_chunks);
```

Comment:
I found one issue from usage

If you trigger an exception which has multiple chunks (probably any message which has multiple chunks), the state will be set to `HITRETURN` which will freeze the UI until Enter is pressed. I discovered this situation by accident when `nvim_get_mode()` evaluated to `"r"` while I was trying to debug the problem.

https://github.com/neovim/neovim/blob/419e0d117d61f6d22f696a8833541dd1691c92fb/src/nvim/state.c#L150-L152

`HITRETURN` should be ignored if the chunks have not been drawn to the built-in message output buffer.

## To reproduce

1.
```
function! Func(...)
endfunction
set msgfunc=Func
```
2. Modify a buffer
3. Try to `:quit`
4. The cursor is stuck in an empty command line until you press Enter, although no message has been drawn

Reply:
The following changes avoid the cursor being stuck on "Press Enter to continue"

```patch
--- a/src/nvim/message.c
+++ b/src/nvim/message.c
@@ -1104,7 +1104,7 @@ void wait_return(int redraw)
 
   // If using ":silent cmd", don't wait for a return.  Also don't set
   // need_wait_return to do it later.
-  if (msg_silent != 0) {
+  if (msg_silent != 0 || msg_enable_msgfunc()) {
     return;
   }
```

Now there's a problem that after exiting the command line, when you come back **the first time**, the command line is invisible. After typing one letter, it appears again. 

https://user-images.githubusercontent.com/17429390/144701987-66b0aef6-7ca3-497c-9a7f-d40bab7e4653.mp4

If you enter, exit, and enter again, that problem does not occur and the command line is visible from the start.

I guess some variable is not being cleared properly due to the early return.

Reply:
OK.  I have fixed.
msgfunc cannot use for hit-enter commands though...